sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/PartitionPump.cs

```diff
@@ -0,0 +1,227 @@
+
+                        InnerConsumer = InnerClient.CreateConsumer(ConsumerGroup, PartitionId, Options.InitialEventPosition);
+
+                        await PartitionProcessor.InitializeAsync().ConfigureAwait(false);
+
+                        RunningTask = RunAsync(RunningTaskTokenSource.Token);
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   Stops the partition pump.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public Task StopAsync() => StopAsync(PartitionProcessorCloseReason.Shutdown);
+
+        /// <summary>
+        ///   Stops the partition pump.  In case it hasn't been started, nothing happens.
+        /// </summary>
+        ///
+        /// <param name="reason">The reason why the partition pump is being closed.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task StopAsync(PartitionProcessorCloseReason reason)
+        {
+            if (RunningTask != null)
+            {
+                await RunningTaskSemaphore.WaitAsync().ConfigureAwait(false);
+
+                try
+                {
+                    if (RunningTask != null)
+                    {
+                        RunningTaskTokenSource.Cancel();
+                        RunningTaskTokenSource = null;
+
+                        try
+                        {
+                            await RunningTask.ConfigureAwait(false);
+                        }
+                        finally
+                        {
+                            RunningTask = null;
+                        }
+
+                        await InnerConsumer.CloseAsync().ConfigureAwait(false);
+                        InnerConsumer = null;
+
+                        await PartitionProcessor.CloseAsync(reason).ConfigureAwait(false);
+                    }
+                }
+                finally
+                {
+                    RunningTaskSemaphore.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        ///   The main loop of a partition pump.  It receives events from the Azure Event Hubs service
+        ///   and delegates their processing to the inner partition processor.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task RunAsync(CancellationToken cancellationToken)
+        {
+            var retryPolicy = new BasicRetryPolicy(new RetryOptions());
```

Comment:
nit:  This could be a readonly static... not that its important enough to make another iteration, but maybe keep an informal note around somewhere for the next time you make tweaks?

Reply:
Noted.